### ISSUE TITLE
RFC: Lazy Kernel Launch

### ISSUE THREAD

**Issue by captainsafia** [nteract/archived-desktop-app#4762]
**Context**

Over the past few years, we've tracked quite a few issues on nteract around improving the kernel launch process. This RFC is meant to capture all of these tracked issues and propose a definitive solution for all of them. Here are some of the issues that have been opened in the past:

* Caching of execution requests during slow kernel starts ( https://github.com/nteract/nteract/issues/1431)
* Buffer cell evaluations while kernel is starting ( https://github.com/nteract/nteract/issues/3368)

Also, @rgbkrk has taken at a stab at implementing buffering for execution requests in the past in this (now-closed) PR: .https://github.com/nteract/nteract/pull/2879.

**Problem**

nteract should support "lazy" kernel launches and improve the kernel launch API to be more robust to kernels that take a long time to launch. This is particular relevant for the following scenarios:

* Users who are using nteract with slow-to start kernels who want to be able to queue cells for execution while their kernel is still starting up.
* Developers who are building nteract-based applications and what to "delay" the launching of a kernel which might be hosted on price-y compute resources (like a GPU host).
* Users who want to open notebooks in view only mode without connecting to a kernel until needed -- similar to the pattern that Google's Colab notebook solution provides.

**Proposed Solution**

nteract should standardize the experience for launching kernels lazily and buffering execution messages as needed. I'd also like to take advantage of this RFC to do some other housekeeping work on the nteract kernel interaction to make it easier to do other things (like support swapping out the execution back-end used to other protocols). I think this work can be achieved by repurposing the `SendExecuteRequest` action to be the action that that execution epics will listen to when determining when to execute code. Currently, the `ExecuteCell` is directly mapped to the execution epics. Instead, I imagine the flow will look like this: 

`ExecuteCell`, `ExecuteFocusedCell`, `ExecuteCellsBelow`, and any other UI-specific execution requests will be mapped to a  `SendExecuteRequest` action.

The `SendExecuteRequest` action will be consumed by an epic that checks to see if a kernel is already available for the ContentRef the execution request is associated with. If so, it will send the action to the `executeCellStream`. If not, it will dispatch an action to launch the kernel and use RxJS's `bufferUntil` operator to buffer the remaining execute requests until the kernel has completely launched.

The interface for the `SendExecuteRequest` action should look something like this.

```javascript
{
	type: "SEND_EXECUTE_REQUEST",
	payload: {
		message: JupyterMessage;
		contentRef: ContentRef;
		id: CellId;
	}
}
```

To implement this, the following work would need to be done.

1. No longer use the `SendExecuteRequest` action in the fashion that it is currently used now (as an internal action to complete follow-on work after an execution has been sent). I've already done this work in  https://github.com/nteract/nteract/pull/4717  so we can repurpose that.
2. Add a set of epics that listen to the various `ExecuteCell` actions and maps them to the new `SendExecuteRequest` action payload.
3. Update the top-level execution epic to listen for the `SendExecuteRequest` and map that to the cell execution Observable.
4. Add logic to the epic from #3 that checks if a kernel exists for the current content and dispatches a launch action that it waits on if it doesn't.

It's important to note that to take full advantage of these changes, consumers of the nteract core SDK should not include the `launchKernelWhenNotebookReady` epic in their application's epic middleware. This epic is designed for those who want to build an experience that mirrors Jupyter classic in that a kernel is immediately launched when the user opens the notebook.

Some considerations:

* Instead of using RxJS operators to buffer the execution requests, should they be stored in a message queue in the Redux store?
* Should the `id` property in the `SendExecuteRequest` be optional? This would give us the flexibility to use nteract's execution API's in non-notebook contexts, per the vision of Kyle's origin PR.
* How do we handle scenarios when the kernel fails to launch? We'll we continue to buffer execution requests indefinitely? We probably need to set a sensible limit on the size of the execution queue?

**Comment by captainsafia** [comment]
Some more thoughts on this after conversation with @languy and others.

> Instead of using RxJS operators to buffer the execution requests, should they be stored in a message queue in the Redux store?

Storing the queued messages in the Redux store is a better option for debugability and monitoring.

```
- core
  - entities
     - content
     - kernels
     ...
     - messages: An Immutable.List of SendExecuteRequests that represents a FIFO queue
```

> Should the id property in the SendExecuteRequest be optional? This would give us the flexibility to use nteract's execution API's in non-notebook contexts, per the vision of Kyle's origin PR.

Yes, it will still be necessary so we don't have to change the existing execute cell stream too much. API consumers are expected to provide a sensible value for this here.

> How do we handle scenarios when the kernel fails to launch? We'll we continue to buffer execution requests indefinitely? We probably need to set a sensible limit on the size of the execution queue?

If the queue is too full, it's sensible to stop accepting requests all together. It doesn't make sense to drop the first few queued actions in order to accommodate newer requests to execute.

We can have another epic that watches the size of the queue by using the `withLatestFrom(state$)` operator and dispatches an error action if the queue is too full. SDK-consumers can listen on this action to dispatch alerts to their users.

We should expose a set of actions for clearing the queue, popping elements out of the queue, and getting the current elements in the queue.

> What to do when the kernel is stuck executing a message?

The existing logic for queuing will be leveraged. If the kernel is launched, we will send the execution requests but the kernel will not respond back until it is ready then we will manage the responses for the rest of the cells.

> With this approach, is there a possibility that we end up with multiple requests to start a kernel?

The logic will check to see if the kernel is either not launched, meaning that we have no `kernel.channels` object or if the `status` of the kernel is in `starting`.

**Comment by captainsafia** [comment]
This shipped to users of the nteract desktop app this week so I will go ahead and close this.

No major bugs with the feature aside from one oversight that we made when implementing this in the desktop app.

It was identified in https://github.com/nteract/nteract/issues/4894. We will need to remove the logic in the desktop app that expects users to have kernelspecs installed since this is technically no longer required when using the desktop app. This logic is implemented in `applications/desktop/src/main/index.ts` in the `fullAppReady$` subscriber.
